// ohm/dist/infra/parse-manager.js
'use strict';

var baseManager = require('./base-manager');
var ParseServer = require('parse-server').ParseServer;
var ParseDashboard = require('parse-dashboard');
var S3Adapter = require('parse-server').S3Adapter;
var parseManager = Object.assign({}, baseManager, {
  configureCommon: function configureCommon(nconf, app, io) {
    app.use(nconf.get('PARSE_PATH'), new ParseServer({
      databaseURI: nconf.get('MONGO_URI') ? nconf.get('MONGO_URI') : 'mongodb://' + nconf.get('MONGO_HOST') + ':' + nconf.get('MONGO_PORT') + '/' + nconf.get('MONGO_DB'),
      appId: nconf.get('PARSE_APPLICATION_ID'),
      masterKey: nconf.get('PARSE_MASTER_KEY'),
      serverURL: nconf.get('PARSE_SERVER_URL')
    }));
  },
  configureDevelopmentEnv: function configureDevelopmentEnv(nconf, app, io) {
    if (nconf.get('PARSE_DASHBOARD')) {
      app.use(nconf.get('PARSE_DASHBOARD'), new ParseDashboard({
        'apps': [{
          'serverURL': nconf.get('PARSE_SERVER_URL'),
          'appId': nconf.get('PARSE_APPLICATION_ID'),
          'masterKey': nconf.get('PARSE_MASTER_KEY'),
          'appName': nconf.get('MONGO_DB')
        }]
      }, false));
    }
  }
});

module.exports = parseManager;
